src/mongo/db/pipeline/expression.cpp

```diff
@@ -2716,31 +2730,102 @@ Value ExpressionIndexOfArray::evaluate(const Document& root) const {
+                                                const ExpressionVector& operands,
+                                                size_t arrayLength) const {
+    std::vector<Value> deps;
+    deps.push_back(operands[1]->evaluate(root));
+
+    int startIndex = 0;
+    if (operands.size() > 2) {
+        Value startIndexArg = operands[2]->evaluate(root);
         uassertIfNotIntegralAndNonNegative(startIndexArg, getOpName(), "starting index");
-        startIndex = static_cast<size_t>(startIndexArg.coerceToInt());
+        startIndex = static_cast<int>(startIndexArg.coerceToInt());
     }
+    deps.push_back(Value(startIndex));
 
-    size_t endIndex = array.size();
-    if (vpOperand.size() > 3) {
-        Value endIndexArg = vpOperand[3]->evaluate(root);
+    int endIndex = arrayLength;
+    if (operands.size() > 3) {
+        Value endIndexArg = operands[3]->evaluate(root);
         uassertIfNotIntegralAndNonNegative(endIndexArg, getOpName(), "ending index");
         // Don't let 'endIndex' exceed the length of the array.
-        endIndex = std::min(array.size(), static_cast<size_t>(endIndexArg.coerceToInt()));
+        endIndex = std::min(arrayLength, static_cast<size_t>(endIndexArg.coerceToInt()));
+    }
+    deps.push_back(Value(endIndex));
+    return deps;
+}
+/**
+ *  This class handles the case where IndexOfArray is given an ExpressionConstant
+ *  instead of using a vector and searching through it we can use a unordered_map
+ *  for O(1) lookup time.
+ **/
+class ExpressionIndexOfArray::Optimized : public ExpressionIndexOfArray {
+public:
+    Optimized(const boost::intrusive_ptr<ExpressionContext>& expCtx,
+              const ValueUnorderedMap<int>& indexMap,
+              const ExpressionVector& operands)
+        : ExpressionIndexOfArray(expCtx), _indexMap(std::move(indexMap)) {
+        vpOperand = operands;
     }
 
-    for (size_t i = startIndex; i < endIndex; i++) {
-        if (getExpressionContext()->getValueComparator().evaluate(array[i] == searchItem)) {
-            return Value(static_cast<int>(i));
+    virtual Value evaluate(const Document& root) const {
+        std::vector<Value> operands = parseDeps(root, vpOperand, _indexMap.size());
+        auto index = _indexMap.find(operands[0]);
+        auto comparator = getExpressionContext()->getValueComparator();
+        Value vIndex = Value(index->second);
+        if (index != _indexMap.end() && comparator.evaluate(Value(index->second) >= operands[1]) &&
+            comparator.evaluate(Value(index->second) < operands[2])) {
+            return Value(index->second);
+        } else {
+            // If the item we are searching is not found or if not found in given range return -1.
+            return Value(-1);
         }
     }
 
-    return Value(-1);
+private:
+    const ValueUnorderedMap<int> _indexMap;
+};
+
+intrusive_ptr<Expression> ExpressionIndexOfArray::optimize() {
+    ExpressionNary::optimize();
```

Comment:
Add a comment here indicating that this will optimize the arguments to this expression. In particular, if all arguments are constant, this should immediately evaluate to a constant. Can you add tests for that like you did for ExpressionArray?